### PR TITLE
Monte-Carlo Tree Search algorithm

### PR DESCRIPTION
**Pull request by emailweixu**

### CODE REVIEW

File: alf/algorithms/mcts_algorithm.py [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+       of the code more difficult and error prone. And it indeed seems there is
+       a bug in the pseudocode related to this. More concretely, in the pseudocode,
+       line 524 suggests that the value_sum is relative to a changing player;
+       line 528 suggests that all the rewards along a path are relative to a same
+       player; while line 499 combines the reward and value without considering the player.
+    2. When calculating UCB score, the pseudocode normalizes value before adding
+       with reward. We normalize after summing reward and value.
+    3. When calculating UCB score, if the visit count of the node is 0, the value
+       component of the score is 0 in the pseudocode. We use normlize(0) instead.
+    4. The pseudocode intilizes the visit count of root to 0. We initialize it
```
Comment (Haichao-Zhang) [inline alf/algorithms/mcts_algorithm.py]: intilizes --> initializes

File: alf/algorithms/mcts_algorithm.py [hunk 2]
```diff
@@ -0,0 +1,496 @@
...
+       player 1 depending on who is on current turn. It makes reasoning the logic
+       of the code more difficult and error prone. And it indeed seems there is
+       a bug in the pseudocode related to this. More concretely, in the pseudocode,
+       line 524 suggests that the value_sum is relative to a changing player;
+       line 528 suggests that all the rewards along a path are relative to a same
+       player; while line 499 combines the reward and value without considering the player.
+    2. When calculating UCB score, the pseudocode normalizes value before adding
+       with reward. We normalize after summing reward and value.
+    3. When calculating UCB score, if the visit count of the node is 0, the value
+       component of the score is 0 in the pseudocode. We use normlize(0) instead.
```
Comment (Haichao-Zhang) [inline alf/algorithms/mcts_algorithm.py]: normlize(0) --> normalize(0)

File: alf/algorithms/mcts_algorithm.py [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+        self.best_child_index = torch.zeros(
+            shape, dtype=torch.int64, device='cpu')
+        self.ucb_score = torch.zeros(shape)
+
+    def update_value_stats(self, values):
+        self.maximum = max(self.maximum, values.max())
+        self.minimum = min(self.minimum, values.min())
+
+    def normalize_value(self, value):
+        if self.maximum > self.minimum:
```
Comment (Haichao-Zhang) [inline alf/algorithms/mcts_algorithm.py:120]: The default values set in line 80 also satisfy this condition (```float('inf') > -float('inf')```), but if we do normalization as in line 123, we might get nan values?

File: alf/algorithms/mcts_algorithm.py [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+            known_value_bounds=(None, None),
+    ):
+        r"""
+        Args:
+            observation_spec (nested TensorSpec): representing the observations.
+            action_spec (nested BoundedTensorSpec): representing the actions.
+            num_simulations (int): the number of simulations per search (calls to model)
+            root_dirichlet_alpha (float): alpha of dirichlet prior for exploration
+            root_exploration_fraction (float): noise generated by the dirichlet distribution
+                is combined with the action distribution from the model to used as
```
Comment (Haichao-Zhang) [inline alf/algorithms/mcts_algorithm.py]: to be used

File: alf/algorithms/mcts_algorithm.py [hunk 5]
```diff
@@ -0,0 +1,496 @@
...
+            discount (float): reward discount factor
+            is_two_player_game (bool): whether this is a two player (zero-sum) game
+            visit_softmax_temperature_fn (Callable): function for calculating the
+                softmax temperature for sampling action based on the visit_counts
+                of the children of the root. :math:`P(a) \propto \exp(visit_count/t)`.
+                This function is called as ``visit_softmax_temperature_fn(steps)``,
+                where ``steps`` is a vector representing the number of steps in
+                the games. And it is expected to return a float vector of the same
+                shape as ``steps``.
+            num_sampled_actions (int): number of sampled action for expanding a node
```
Comment (Haichao-Zhang) [inline alf/algorithms/mcts_algorithm.py]: ```num_sampled_actions```: currently not included in the argument list of ```__init__```

File: alf/algorithms/mcts_algorithm.py [hunk 6]
```diff
@@ -0,0 +1,496 @@
...
+        self._model = model
+
+    def predict_step(self, time_step: TimeStep, state: MCTSState):
+        """Predict the action.
+
+        Args:
+            time_step (TimeStep): the time step structure
+            state (MCTSState): state of the algorithm
+        Returns:
+            AlgOutput:
```
Comment (Haichao-Zhang) [inline alf/algorithms/mcts_algorithm.py]: AlgStep

File: alf/algorithms/mcts_algorithm.py [hunk 7]
```diff
@@ -0,0 +1,496 @@
...
+        if self.maximum > self.minimum:
+            # We normalize only when we have set the maximum and minimum values.
+            return (value - self.minimum) / (self.maximum - self.minimum)
+        return value
+
+    def calc_value(self, nodes):
+        return self.value_sum[nodes] / (self.visit_count[nodes] + 1e-30)
+
+
+MAXIMUM_FLOAT_VALUE = float('inf')
```
Comment (Haichao-Zhang) [inline alf/algorithms/mcts_algorithm.py]: already defined in line69

Comment (emailweixu) [reply alf/algorithms/mcts_algorithm.py]: removed.

File: alf/algorithms/mcts_algorithm.py (hunk 3 above)
Comment (emailweixu) [reply alf/algorithms/mcts_algorithm.py:120]: fixed

File: alf/algorithms/mcts_algorithm.py (hunk 1 above)
Comment (emailweixu) [reply alf/algorithms/mcts_algorithm.py]: fixed.

File: alf/algorithms/mcts_algorithm.py (hunk 2 above)
Comment (emailweixu) [reply alf/algorithms/mcts_algorithm.py]: changed to 0.5 in the subsequent commit

File: alf/algorithms/mcts_algorithm.py (hunk 4 above)
Comment (emailweixu) [reply alf/algorithms/mcts_algorithm.py]: fixed.

File: alf/algorithms/mcts_algorithm.py (hunk 5 above)
Comment (emailweixu) [reply alf/algorithms/mcts_algorithm.py]: removed

File: alf/algorithms/mcts_algorithm.py (hunk 6 above)
Comment (emailweixu) [reply alf/algorithms/mcts_algorithm.py]: fixed.

File: alf/algorithms/mcts_algorithm.py (hunk 3 above)
Comment (Haichao-Zhang) [reply alf/algorithms/mcts_algorithm.py:120]: The initialization is correct. I didn't notice the flipped sign.

**Review by Haichao-Zhang — APPROVED** [review]
